server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java

```diff
@@ -0,0 +1,473 @@
+     * period that needs to lapse before the current recommendation should be applied.
+     * <p>
+     * If auto sharding is not applicable for a data stream (e.g. due to {@link #DATA_STREAMS_AUTO_SHARDING_EXCLUDES_SETTING}) the target
+     * number of shards will be 0 and cool down remaining {@link TimeValue#MAX_VALUE}.
+     */
+    public record AutoShardingResult(
+        AutoShardingType type,
+        int currentNumberOfShards,
+        int targetNumberOfShards,
+        TimeValue coolDownRemaining,
+        @Nullable Double writeLoad
+    ) implements Writeable, ToXContentObject {
+
+        public static final ParseField AUTO_SHARDING_TYPE = new ParseField("type");
+        public static final ParseField CURRENT_NUMBER_OF_SHARDS = new ParseField("current_number_of_shards");
+        public static final ParseField TARGET_NUMBER_OF_SHARDS = new ParseField("target_number_of_shards");
+        public static final ParseField COOLDOWN_REMAINING = new ParseField("cool_down_remaining");
+        public static final ParseField WRITE_LOAD = new ParseField("write_load");
+
+        public AutoShardingResult(
+            AutoShardingType type,
+            int currentNumberOfShards,
+            int targetNumberOfShards,
+            TimeValue coolDownRemaining,
+            @Nullable Double writeLoad
+        ) {
+            this.type = type;
+            this.currentNumberOfShards = currentNumberOfShards;
+            this.targetNumberOfShards = targetNumberOfShards;
+            this.coolDownRemaining = coolDownRemaining;
+            this.writeLoad = writeLoad;
+        }
+
+        public AutoShardingResult(StreamInput in) throws IOException {
+            this(in.readEnum(AutoShardingType.class), in.readVInt(), in.readVInt(), in.readTimeValue(), in.readOptionalDouble());
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.field(AUTO_SHARDING_TYPE.getPreferredName(), type);
+            builder.field(CURRENT_NUMBER_OF_SHARDS.getPreferredName(), currentNumberOfShards);
+            builder.field(TARGET_NUMBER_OF_SHARDS.getPreferredName(), targetNumberOfShards);
+            builder.field(COOLDOWN_REMAINING.getPreferredName(), coolDownRemaining.toHumanReadableString(2));
+            builder.field(WRITE_LOAD.getPreferredName(), writeLoad);
+            builder.endObject();
+            return builder;
+        }
+
+        @Override
+        public String toString() {
+            return "{ type: "
+                + type
+                + ", currentNumberOfShards: "
+                + currentNumberOfShards
+                + ", targetNumberOfShards: "
+                + targetNumberOfShards
+                + ", coolDownRemaining: "
+                + coolDownRemaining
+                + ", writeLoad: "
+                + writeLoad
+                + " }";
```

Comment:
I'll work on something different in the subsequent PR (this will be removed). I added this because a rollover `Condition` uses `toString` as the representation for the `RolloverResponse` and cannot be overridden https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/admin/indices/rollover/Condition.java#L68 

Will do something different in the next PR that handles the rollover.